tests/FSharp.Test.Utilities/TestFramework.fs

```diff
@@ -10,28 +10,59 @@ open Scripting
-let tryCreateTemporaryDirectory () =
-    let date() = DateTime.Now.ToString("yyyy-MM-dd")
-    let now() = $"{date()}-{Guid.NewGuid().ToString()}"
-    let directory = Path.Combine(Path.GetTempPath(), now()).Replace('-', '_')
-    Directory.CreateDirectory(directory).FullName
-
-// Create a temporaryFileName -- newGuid is random --- there is no point validating the file already exists because: threading and Path.ChangeExtension() is commonly used after this API
-let tryCreateTemporaryFileName () =
-    let directory = tryCreateTemporaryDirectory ()
-    let fileName = ("Temp-" + Guid.NewGuid().ToString() + ".tmp").Replace('-', '_')
-    let filePath = Path.Combine(directory, fileName)
-    filePath
-
-// Create a temporaryFileName -- newGuid is random --- there is no point validating the file already exists because: threading and Path.ChangeExtension() is commonly used after this API
-let tryCreateTemporaryFileNameInDirectory (directory: DirectoryInfo) =
-    let fileName = ("Temp-" + Guid.NewGuid().ToString() + ".tmp").Replace('-', '_')
-    let filePath = Path.Combine(directory.FullName, fileName)
-    filePath
+let getShortId() = Guid.NewGuid().ToString().[..7]
+
+// Temporary directory is TempPath + "/FSharp.Test.Utilities/yyy-MM-dd-xxxxxxx/"
+let tempDirectoryOfThisTestRun =
+    let tempDir = Path.GetTempPath()
+    let today = DateTime.Now.ToString("yyyy-MM-dd")
+    DirectoryInfo(tempDir)
+        .CreateSubdirectory($"FSharp.Test.Utilities/{today}-{getShortId()}")
+        .FullName
+
+let createTemporaryDirectory (part: string) =
+    DirectoryInfo(tempDirectoryOfThisTestRun)
+        .CreateSubdirectory($"{part}-{getShortId()}")
+        .FullName
+
+let getTemporaryFileName () =
+    (createTemporaryDirectory "temp") ++ $"tmp_{getShortId()}"
+
+let getTemporaryFileNameInDirectory (directory: string) =
+    directory ++ $"tmp_{getShortId()}"
+
+// Well, this function is AI generated.
+let rec copyDirectory (sourceDir: string) (destinationDir: string) (recursive: bool) =
+    // Get information about the source directory
+    let dir = DirectoryInfo(sourceDir)
+
+    // Check if the source directory exists
+    if not dir.Exists then
+        raise (DirectoryNotFoundException($"Source directory not found: {dir.FullName}"))
+
+    // Create the destination directory
+    Directory.CreateDirectory(destinationDir) |> ignore
+
+    // Get the files in the source directory and copy to the destination directory
+    for file in dir.EnumerateFiles() do
+        let targetFilePath = Path.Combine(destinationDir, file.Name)
+        file.CopyTo(targetFilePath) |> ignore
+
+    // If recursive and copying subdirectories, recursively call this method
+    if recursive then
+        for subDir in dir.EnumerateDirectories() do
+            let newDestinationDir = Path.Combine(destinationDir, subDir.Name)
+            copyDirectory subDir.FullName newDestinationDir true
+
+let copyTestDirectoryToTempLocation source (destinationSubDir: string) =
```

Comment:
Help me understand this last thing - why is it needed? As in, why cannot we create some transient stuff in the temp location right away?

Reply:
This creates a two dirs deep structure.
Some tests (typeproviders) tend to reach up to the dir above of their nominal config directory.
This, in fact should be a special case as we don't need it in general.

Reply:
Right, type providers again. Yeah maybe if we can narrow down their own voodoo to their place, it would be helpful. Ideally with a note to rewrite all that properly one day :D